File: src/configuration.ts
import { IS_CLIENT } from "./internal";
import NetServerEvent from "./ServerEvent";
import NetServerFunction from "./ServerFunction";
import NetServerAsyncFunction from "./ServerAsyncFunction";

const runService = game.GetService("RunService");

const IS_SERVER = runService.IsServer();

type InvalidPropTypesHandler = (
	who: NetServerEvent | NetServerFunction | NetServerAsyncFunction,
	srcPlayer: Player,
) => void;

interface RbxNetConfigItem {
	/**
	 * The throttle reset timer (default: 60 seconds)
	 */
	ServerThrottleResetTimer: number;

	EnableDebugMessages: boolean;

	InvalidPropTypesHandler?: InvalidPropTypesHandler;

	/**
	 * The message shown when the throttle has been exceeded.
	 * {player} will be replaced with the player's name!
	 */
	ServerThrottleMessage: string;

	/** @internal */
	__stfuTypescript: undefined;
}

let throttleResetTimer = 60;
let rateLimitReachedMessage = "Request limit exceeded ({limit}) by {player} via {remote}";
namespace NetConfig {
	/** @internal */
	export let DebugEnabled = false;

	/** @internal */
	export let InvalidPropTypesHandlerFunc: InvalidPropTypesHandler | undefined;

	/** @rbxts client */
	export function SetClientConfiguration<K extends keyof Pick<RbxNetConfigItem, "EnableDebugMessages">>(
		key: K,
		value: RbxNetConfigItem[K],
	): void {
		assert(IS_CLIENT, "Use SetConfiguration on the server!");
		if (key === "EnableDebugMessages") {
			DebugEnabled = value as boolean;
		}
	}

	/** @rbxts server */
	export function SetConfiguration<K extends keyof RbxNetConfigItem>(key: K, value: RbxNetConfigItem[K]) {
		assert(IS_SERVER, "Cannot set configuration on client!");
		if (key === "ServerThrottleResetTimer") {
			throttleResetTimer = value as number;
		} else if (key === "ServerThrottleMessage") {
			rateLimitReachedMessage = value as string;
		} else if (key === "EnableDebugMessages") {
			DebugEnabled = value as boolean;
		} else if (key === "InvalidPropTypesHandler") {
			InvalidPropTypesHandlerFunc = value as InvalidPropTypesHandler;
		}
	}

	export function GetConfiguration<K extends keyof RbxNetConfigItem>(key: K): RbxNetConfigItem[K] {
		if (key === "ServerThrottleResetTimer") {
			assert(IS_SERVER, "ServerThrottleResetTimer is not used on the client!");
			return throttleResetTimer as RbxNetConfigItem[K];
		} else if (key === "ServerThrottleMessage") {
			assert(IS_SERVER, "ServerThrottleMessage is not used on the client!");
			return rateLimitReachedMessage as RbxNetConfigItem[K];
		} else if (key === "EnableDebugMessages") {
			return DebugEnabled as RbxNetConfigItem[K];
		} else {
			return undefined as RbxNetConfigItem[K];
		}
	}

	/** @internal */
	export function DebugWarn(...message: Array<unknown>) {
		if (DebugEnabled) {
			warn("[rbx-net-debug]", ...message);
		}
	}

	/** @internal */
	export function DebugLog(...message: Array<unknown>) {
		if (DebugEnabled) {
			print("[rbx-net-debug]", ...message);
		}
	}
}

export = NetConfig;

File: configuration.lua



local TS = require(script.Parent.vendor.RuntimeLib);
local exports;
local NetConfig;
local IS_CLIENT = TS.import(script, script.Parent, "internal").IS_CLIENT;
local runService = game:GetService("RunService");
local IS_SERVER = runService:IsServer();
local throttleResetTimer = 60;
local rateLimitReachedMessage = "Request limit exceeded ({limit}) by {player} via {remote}";
NetConfig = NetConfig or {} do
	local _0 = NetConfig;
	_0.DebugEnabled = false;
	local function SetClientConfiguration(key, value)
		assert(IS_CLIENT, "Use SetConfiguration on the server!");
		if key == "EnableDebugMessages" then
			_0.DebugEnabled = value;
		end;
	end;
	local function SetConfiguration(key, value)
		assert(IS_SERVER, "Cannot set configuration on client!");
		if key == "ServerThrottleResetTimer" then
			throttleResetTimer = value;
		elseif key == "ServerThrottleMessage" then
			rateLimitReachedMessage = value;
		elseif key == "EnableDebugMessages" then
			_0.DebugEnabled = value;
		elseif key == "InvalidPropTypesHandler" then
			_0.InvalidPropTypesHandlerFunc = value;
		end;
	end;
	local function GetConfiguration(key)
		if key == "ServerThrottleResetTimer" then
			assert(IS_SERVER, "ServerThrottleResetTimer is not used on the client!");
			return throttleResetTimer;
		elseif key == "ServerThrottleMessage" then
			assert(IS_SERVER, "ServerThrottleMessage is not used on the client!");
			return rateLimitReachedMessage;
		elseif key == "EnableDebugMessages" then
			return _0.DebugEnabled;
		else
			return nil;
		end;
	end;
	local function DebugWarn(...)
		local message = { ... };
		if _0.DebugEnabled then
			warn("[rbx-net-debug]", unpack(message));
		end;
	end;
	local function DebugLog(...)
		local message = { ... };
		if _0.DebugEnabled then
			print("[rbx-net-debug]", unpack(message));
		end;
	end;
	_0.SetClientConfiguration = SetClientConfiguration;
	_0.SetConfiguration = SetConfiguration;
	_0.GetConfiguration = GetConfiguration;
	_0.DebugWarn = DebugWarn;
	_0.DebugLog = DebugLog;
end;
exports = NetConfig;
return exports;
